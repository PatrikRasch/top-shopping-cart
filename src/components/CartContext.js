import React, { useContext, useState, useEffect } from "react"; // The basic imports

const CartContext = React.createContext(); // Creates overall context

export function useCart() {
  return useContext(CartContext);
}

export function CartProvider({ children }) {
  const [numOfItemsInCart, setNumOfItemsInCart] = useState(0);
  const [itemsInCart, setItemsInCart] = useState([]);
  const [subtotalPrice, setSubtotalPrice] = useState(0);

  const calculateTotalPriceItem = (array, index) => {
    const priceForOne = array[index].price; // Declare fixed price
    const totalOfItemInCart = array[index].numOfItemInCart; // Declare number of item in cart
    array[index].priceForAll = Math.floor(priceForOne * totalOfItemInCart * 100) / 100; // Calculate total price for all
  };

  const addItemToCart = (item) => {
    let duplicate = null; // Variable to match itemsInCart item
    // Check if the item already exists in the array
    const itemAlreadyExists = itemsInCart.some((element) => {
      if (item.title === element.title) {
        duplicate = item;
        // Find the item in the array that is to be updated
        const toBeUpdated = itemsInCart.find((item) => item.title === duplicate.title);
        const duplicatedArray = [...itemsInCart]; // Make duplicate of itemsInCart
        const indexToBeUpdated = duplicatedArray.indexOf(toBeUpdated); // Find the index to update in itemsInCart
        duplicatedArray[indexToBeUpdated].numOfItemInCart += 1; // Add +1 to the numOfItemInCart to the item
        calculateTotalPriceItem(duplicatedArray, indexToBeUpdated);
        setItemsInCart(duplicatedArray); // Update state
        return true; // Item exists
      }
      return false; // Item doesn't exist
    });
    if (itemAlreadyExists) return;
    else {
      setItemsInCart((current) => [...current, item]);
    } // Runs only when item doesn't exist â†’ adds new item to state
  };

  /// Update the subtotal price the user has to pay
  useEffect(() => {
    const initialValue = 0;
    const newSubtotalPrice = itemsInCart.reduce(
      (acc, currentValue) => currentValue.priceForAll + acc,
      initialValue
    );
    const newSubtotalPriceRounded = Math.floor(newSubtotalPrice * 100) / 100;
    setSubtotalPrice(newSubtotalPriceRounded);
  }, [itemsInCart, numOfItemsInCart]);

  /// Updates the total number of items in the cart
  const updateTotalNumOfItemsInCart = () => {
    const initialValue = 0;
    const totalItemsInCart = itemsInCart.reduce(
      (acc, currentValue) => currentValue.numOfItemInCart + acc,
      initialValue
    );
    setNumOfItemsInCart(totalItemsInCart);
  };

  /// Add or remove an extra of item that already exists in cart
  const addOrRemoveOneOfItem = (item, plusOrMinus) => {
    const duplicatedArray = [...itemsInCart];
    const indexOfItem = itemsInCart.indexOf(item);
    duplicatedArray[indexOfItem].numOfItemInCart += plusOrMinus;
    calculateTotalPriceItem(duplicatedArray, indexOfItem);
    setItemsInCart(duplicatedArray);
    updateTotalNumOfItemsInCart();
    removeItemFromCart();
  };

  const addToCartTotal = () => {
    setNumOfItemsInCart(numOfItemsInCart + 1);
  };

  /// Removes items from the cart if the number of the item in the cart is 0
  const removeItemFromCart = () => {
    itemsInCart.forEach((item) => {
      if (item.numOfItemInCart === 0) {
        const duplicatedArray = [...itemsInCart];
        const itemRemovedArray = duplicatedArray.filter((element) => {
          return element !== item;
        });
        setItemsInCart(itemRemovedArray);
      }
    });
  };

  const resetCart = () => {
    setItemsInCart([]);
    setNumOfItemsInCart(0);
  };

  const cartState = {
    itemsInCart,
    numOfItemsInCart,
    subtotalPrice,
    addToCartTotal,
    addItemToCart,
    addOrRemoveOneOfItem,
    resetCart,
  };

  return <CartContext.Provider value={cartState}>{children}</CartContext.Provider>;
}
