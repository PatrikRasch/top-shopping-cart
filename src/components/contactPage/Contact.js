import "./contact.css";

import github from "../../images/contact-page/github-black.webp";
import facebook from "../../images/contact-page/facebook-black.webp";
import messenger from "../../images/contact-page/messenger-black.webp";
import twitter from "../../images/contact-page/twitter-black.webp";
import instagram from "../../images/contact-page/instagram-black.webp";
import footerImage from "../../images/contact-page/footer-orange1.webp";

const Contact = () => {
  return (
    <div className="contact-wrapper">
      <div className="contact-content">
        <div className="all-icons-wrapper">
          <div className="icon-wrapper icon-wrapper-github">
            <a href="https://github.com/PatrikRasch">
              <img src={github} alt="" className="icon" />
            </a>
          </div>

          <div className="icon-wrapper icon-wrapper-facebook">
            <a href="https://www.youtube.com/watch?v=OZshCpFSgwI">
              <img src={facebook} alt="" className="icon" />
            </a>
          </div>
          <div className="icon-wrapper icon-wrapper-messenger">
            <a href="https://www.youtube.com/watch?v=xua4P0MFIgQ">
              <img src={messenger} alt="" className="icon" />
            </a>
          </div>
          <div className="icon-wrapper icon-wrapper-twitter">
            <a href="https://www.youtube.com/watch?v=F4dGW0jHbuo">
              <img src={twitter} alt="" className="icon" />
            </a>
          </div>
          <div className="icon-wrapper icon-wrapper-instagram">
            <a href="https://www.youtube.com/watch?v=wT7oWz40Q1M">
              <img src={instagram} alt="" className="icon" />
            </a>
          </div>
        </div>
        <div className="address-wrapper">
          <div className="address-line1">Unreal Clothing</div>
          <div className="address-line2">Klostergata 90, 7030 Trondheim Norway</div>
        </div>
      </div>
      <div className="footer">
        <img src={footerImage} alt="" className="contact-page-footer" />
      </div>
    </div>
  );
};

export default Contact;
