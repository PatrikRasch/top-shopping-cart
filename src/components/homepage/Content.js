import "./content.css";
import { useNavigate } from "react-router-dom";
// import contentPage1 from "../../images/content-page/content-page1.webp";
// import contentPage2 from "../../images/content-page/content-page2.webp";
// import contentPage3 from "../../images/content-page/content-page3.webp";
import contentPage4 from "../../images/content-page/content-page4.webp";
// import contentPage5 from "../../images/content-page/content-page5.webp";
// import contentPage6 from "../../images/content-page/content-page6.webp";
import contentPage7 from "../../images/content-page/content-page7.webp";
// import contentPage8 from "../../images/content-page/content-page8.webp";
import contentPage9 from "../../images/content-page/content-page9.webp";

const Content = () => {
  const navigate = useNavigate();
  return (
    <div className="content-wrapper">
      <div className="content-images-wrapper">
        <div className="single-image-wrapper">
          <div className="content-image-shading"></div>
          <img src={contentPage9} alt="" className="content-image content-image1" />
        </div>
        <div className="single-image-wrapper">
          <div className="content-image-shading"></div>
          <img src={contentPage7} alt="" className="content-image content-image2" />
        </div>
        <div className="single-image-wrapper">
          <div className="content-image-shading"></div>
          <img src={contentPage4} alt="" className="content-image content-image3" />
        </div>
      </div>
      <div className="content-title">
        <i>Your Gateway</i> into <b>Fashion</b>
      </div>

      <button className="shop-now" onClick={() => navigate("/products")}>
        SHOP NOW
      </button>
    </div>
  );
};

export default Content;
