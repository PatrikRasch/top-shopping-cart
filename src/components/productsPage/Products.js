import "./products.css";
import Product from "./product";

import shirt1 from "./../../images/clothing/shirt1.png";
import shirt2 from "./../../images/clothing/shirt2.png";
import shirt3 from "./../../images/clothing/shirt3.png";
import shirt4 from "./../../images/clothing/shirt4.png";
import shirt5 from "./../../images/clothing/shirt5.png";
import shirt6 from "./../../images/clothing/shirt6.png";
import shirt7 from "./../../images/clothing/shirt7.png";
import shirt11 from "./../../images/clothing/shirt11.png";
import shirt15 from "./../../images/clothing/shirt15.png";
import shirt16 from "./../../images/clothing/shirt16.png";
import tshirt1 from "./../../images/clothing/tshirt1.png";
import tshirt2 from "./../../images/clothing/tshirt2.png";
import jacket1 from "./../../images/clothing/jacket1.png";
import jacket3 from "./../../images/clothing/jacket3.png";
import jacket4 from "./../../images/clothing/jacket4.png";
import hoodie1 from "./../../images/clothing/hoodie1.png";
import polo1 from "./../../images/clothing/polo1.png";
import shoes1 from "./../../images/clothing/shoes1.png";
import shoes2 from "./../../images/clothing/shoes2.png";
import shoes3 from "./../../images/clothing/shoes3.png";

const Products = () => {
  return (
    <div>
      <div className="all-products-background">
        <div className="all-products">
          <Product image={shirt7} title={"White Cotton Polo Shirt"} price={159.99} />
          <Product image={shirt2} title={"Light Blue Signature Twill Dress Shirt "} price={79.99} />
          <Product image={jacket4} title={"Black Leather Jacket"} price={99.99} />
          <Product image={shirt1} title={"White Four-Way Stretch Dress Shirt"} price={99.99} />
          <Product image={shoes3} title={"Black Formal Leather Shoes"} price={99.5} />
          <Product image={polo1} title={"Black Cotton Polo Shirt"} price={99.99} />
          <Product image={shirt3} title={"Gray Business Casual Stretch Shirt"} price={49.99} />
          <Product image={shoes1} title={"Brown Formal Leather Shoes"} price={99.99} />
          <Product image={tshirt1} title={"Grey Premium Stretch T-shirt"} price={99.5} />
          <Product image={shirt6} title={"Black Premium Stretch Dress Shirt"} price={99.99} />
          <Product image={shirt5} title={"White Linen Shirt - Short Sleeve"} price={109.5} />
          <Product image={shirt11} title={"Blue Wrinkle Free Cotton Linen Shirt"} price={69.5} />
          <Product image={hoodie1} title={"Black Casual Hoodie"} price={99.99} />
          <Product image={shirt15} title={"Shirt Bundle of 6"} price={99} />
          <Product image={shirt16} title={"Light Blue Denim Shirt"} price={99.99} />
          <Product image={shoes2} title={"Blue Business Casual Fabric Shoes"} price={99.99} />
          <Product image={tshirt2} title={"Black Premium Cotton T-shirt"} price={99.99} />
          <Product image={jacket1} title={"Green Winter Casual Overcoat Jacket"} price={99.99} />
          <Product image={shirt4} title={"White Stretch Twill Shirt"} price={79.99} />
          <Product image={jacket3} title={"Black Formal Button Up Jacket"} price={99.5} />
        </div>
      </div>
    </div>
  );
};

export default Products;
