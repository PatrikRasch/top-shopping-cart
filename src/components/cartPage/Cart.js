import "./cart.css";
import { useCart } from "../CartContext";

const Cart = (props) => {
  const { itemsInCart, addOrRemoveOneOfItem, subtotalPrice, resetCart } = useCart();

  const checkout = () => {
    if (itemsInCart.length === 0) {
      return alert("Your cart is empty");
    }
    alert("Thank you for checking out my project");
    props.handleCloseCart();
    setTimeout(() => {
      resetCart();
    }, 600);
  };

  const emptyBasketMessage = () => {
    return (
      <div className="empty-basket-message">
        Your basket is empty, go add some unreal clothing to it!
      </div>
    );
  };
  const displayCartItems = () => {
    if (itemsInCart.length === 0) return;
    return itemsInCart.map((item) => (
      <div className="cart-card">
        <div className="cart-card-left">
          <img src={item.image} alt="" className="cart-item-image" />
        </div>
        <div className="cart-card-right">
          <div className="cart-item-title">{item.title}</div>
          <div className="cart-item-price">${item.priceForAll}</div>
          <div className="cart-number-of-item-wrapper">
            <button className="remove-one-item" onClick={() => addOrRemoveOneOfItem(item, -1)}>
              -
            </button>
            <div className="cart-number-of-item">{item.numOfItemInCart}</div>
            <button className="add-one-item" onClick={() => addOrRemoveOneOfItem(item, 1)}>
              +
            </button>
          </div>
        </div>
      </div>
    ));
  };

  return (
    <div className="cart">
      <div className="cart-title">Your Basket</div>
      <div className="cart-all-cards">
        {itemsInCart.length === 0 ? emptyBasketMessage() : displayCartItems()}
      </div>
      <div className="cart-summary">
        <div className="totalPrice">Total: ${subtotalPrice}</div>
        <button className="checkout-button" onClick={() => checkout()}>
          Checkout
        </button>
        <button className="close-button" onClick={() => props.handleCloseCart()}>
          Close
        </button>
      </div>
    </div>
  );
};

export default Cart;
