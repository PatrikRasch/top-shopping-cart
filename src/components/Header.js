import "./header.css";
import shoppingCart from "../images/shopping-cart.png";
import { useNavigate } from "react-router-dom";
import { useCart } from "./CartContext";
import { useState } from "react";
import Cart from "./cartPage/Cart";

const Header = () => {
  const { numOfItemsInCart } = useCart();
  const navigate = useNavigate();
  const [cartShow, setCartShow] = useState(false);
  const [hideCartOverlay, setHideCartOverlay] = useState(true);
  const [timeoutId, setTimeoutId] = useState(null);

  const handleOpenCart = () => {
    setCartShow(true);
    setHideCartOverlay(false);
    clearTimeout(timeoutId);
  };

  const handleCloseCart = () => {
    setCartShow(false);
    const newTimeoutId = setTimeout(() => {
      setHideCartOverlay(true);
    }, 1000);
    setTimeoutId(newTimeoutId);
  };

  return (
    <div>
      <div className="cart-overlay">
        <div
          className={`cart-overlay-cart ${
            cartShow ? "cart-overlay-cart-visible" : "cart-overlay-cart-invisible"
          }`}
        >
          <Cart handleCloseCart={handleCloseCart} />
        </div>
        <div
          className={`cart-overlay-background ${
            cartShow ? "cart-overlay-background-visible" : "cart-overlay-background-invisible"
          } ${hideCartOverlay ? "cart-background-disabled" : ""}`}
          onClick={() => handleCloseCart()}
        ></div>
      </div>
      <div className="header-wrapper-background">
        <div className="header-wrapper">
          <div className="header-title" onClick={() => navigate("/")}>
            UNREAL CLOTHING
          </div>
          <div className="spacer"></div>
          <div className="header-buttons">
            <button className="header-button home-button" onClick={() => navigate("/")}>
              Home
            </button>
            <button className="header-button products-button" onClick={() => navigate("/products")}>
              Products
            </button>
            <button className="header-button contact-button" onClick={() => navigate("/contact")}>
              Contact
            </button>
            <button
              className="header-button cart-button"
              onClick={() => {
                handleOpenCart();
              }}
            >
              <div className="shopping-cart-circle">
                <img
                  src={shoppingCart}
                  alt="Shopping cart button. Go to cart."
                  className="shopping-cart-logo"
                />
                <div className={`item-count ${numOfItemsInCart === 0 ? "hidden" : ""}`}>
                  {numOfItemsInCart}
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Header;
