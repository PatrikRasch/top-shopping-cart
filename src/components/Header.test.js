import React, { useContext } from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Header from "./Header";
import Content from "./homepage/Content";
import Contact from "./contactPage/Contact";
import Products from "./productsPage/Products";
import Cart from "./cartPage/Cart";
import { CartProvider, useCart } from "./CartContext";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import "@testing-library/jest-dom";

const renderSetup = () => {
  render(
    <MemoryRouter>
      <CartProvider>
        <Header />
      </CartProvider>
    </MemoryRouter>
  );
};

describe("Header component", () => {
  it("renders UNREAL CLOTHING", () => {
    renderSetup();
    const headerTitle = screen.getByText("UNREAL CLOTHING");
    expect(headerTitle).toBeInTheDocument();
    expect(headerTitle).toBeVisible();
  });

  it("renders the home button", () => {
    renderSetup();
    const homeButton = screen.getByRole("button", { name: "Home" });
    expect(homeButton).toBeInTheDocument();
    expect(homeButton).toBeVisible();
  });

  it("renders the products button", () => {
    renderSetup();
    const productsButton = screen.getByRole("button", { name: "Products" });
    expect(productsButton).toBeInTheDocument();
    expect(productsButton).toBeVisible();
  });

  it("renders the contact button", () => {
    renderSetup();
    const contactButton = screen.getByRole("button", { name: "Contact" });
    expect(contactButton).toBeInTheDocument();
    expect(contactButton).toBeVisible();
  });

  it("changes to homepage when home button is clicked", () => {
    render(
      <MemoryRouter initialEntries={["/contact"]}>
        <CartProvider>
          <Header />
          <Routes>
            <Route path="/contact" element={<Contact />} />
            <Route path="/" element={<Content />} />
          </Routes>
        </CartProvider>
      </MemoryRouter>
    );

    const homeButton = screen.getByRole("button", { name: "Home" });
    fireEvent.click(homeButton);
    expect(screen.getByText("SHOP NOW")).toBeInTheDocument();
  });

  // Snapshot test
  it("changes to homepage when home button is clicked, snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/contact"]}>
        <CartProvider>
          <Header />
          <Routes>
            <Route path="/contact" element={<Contact />} />
            <Route path="/" element={<Content />} />
          </Routes>
        </CartProvider>
      </MemoryRouter>
    );
    const homeButton = screen.getByRole("button", { name: "Home" });
    fireEvent.click(homeButton);
    expect(asFragment()).toMatchSnapshot();
  });

  it("changes to products page when product button is clicked", () => {
    render(
      <MemoryRouter initialEntries={["/contact"]}>
        <CartProvider>
          <Header />
          <Routes>
            <Route path="/products" element={<Products />} />
            <Route path="/contact" element={<Contact />} />
          </Routes>
        </CartProvider>
      </MemoryRouter>
    );
    const productButton = screen.getByRole("button", { name: "Products" });
    fireEvent.click(productButton);
    expect(screen.getByText("White Cotton Polo Shirt")).toBeInTheDocument();
  });

  it("changes to contact page when contact button is clicked", () => {
    render(
      <MemoryRouter initialEntries={["/products"]}>
        <CartProvider>
          <Header />
          <Routes>
            <Route path="/products" element={<Products />} />
            <Route path="/contact" element={<Contact />} />
          </Routes>
        </CartProvider>
      </MemoryRouter>
    );
    const contactButton = screen.getByRole("button", { name: "Contact" });
    fireEvent.click(contactButton);
    expect(screen.getByText("Unreal Clothing")).toBeInTheDocument();
  });
});
