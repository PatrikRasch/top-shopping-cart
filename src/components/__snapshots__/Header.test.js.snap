// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component changes to homepage when home button is clicked, snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="cart-overlay"
    >
      <div
        class="cart-overlay-cart cart-overlay-cart-invisible"
      >
        <div
          class="cart"
        >
          <div
            class="cart-title"
          >
            Your Basket
          </div>
          <div
            class="cart-all-cards"
          >
            <div
              class="empty-basket-message"
            >
              Your basket is empty, go add some unreal clothing to it!
            </div>
          </div>
          <div
            class="cart-summary"
          >
            <div
              class="totalPrice"
            >
              Total: $0
            </div>
            <button
              class="checkout-button"
            >
              Checkout
            </button>
            <button
              class="close-button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
      <div
        class="cart-overlay-background cart-overlay-background-invisible cart-background-disabled"
      />
    </div>
    <div
      class="header-wrapper-background"
    >
      <div
        class="header-wrapper"
      >
        <div
          class="header-title"
        >
          UNREAL CLOTHING
        </div>
        <div
          class="spacer"
        />
        <div
          class="header-buttons"
        >
          <button
            class="header-button home-button"
          >
            Home
          </button>
          <button
            class="header-button products-button"
          >
            Products
          </button>
          <button
            class="header-button contact-button"
          >
            Contact
          </button>
          <button
            class="header-button cart-button"
          >
            <div
              class="shopping-cart-circle"
            >
              <img
                alt="Shopping cart button. Go to cart."
                class="shopping-cart-logo"
                src="shopping-cart.png"
              />
              <div
                class="item-count hidden"
                data-testid="itemCount"
              >
                0
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="content-wrapper"
  >
    <div
      class="content-images-wrapper"
    >
      <div
        class="single-image-wrapper"
      >
        <div
          class="content-image-shading"
        />
        <img
          alt="background display 1"
          class="content-image content-image1"
          src="content-page9.webp"
        />
      </div>
      <div
        class="single-image-wrapper"
      >
        <div
          class="content-image-shading"
        />
        <img
          alt=""
          class="content-image content-image2"
          src="content-page7.webp"
        />
      </div>
      <div
        class="single-image-wrapper"
      >
        <div
          class="content-image-shading"
        />
        <img
          alt=""
          class="content-image content-image3"
          src="content-page4.webp"
        />
      </div>
    </div>
    <div
      class="content-title"
    >
      <i>
        Your Gateway
      </i>
       into 
      <b>
        Fashion
      </b>
    </div>
    <button
      class="shop-now"
    >
      SHOP NOW
    </button>
  </div>
</DocumentFragment>
`;
